README.md
# Consul

This is docker image to run the [Consul](https://www.consul.io/).

[![](https://images.microbadger.com/badges/image/imega/consul.svg)](http://microbadger.com/images/imega/consul "Get your own image badge on microbadger.com") [![CircleCI](https://circleci.com/gh/imega-docker/consul.svg?style=svg)](https://circleci.com/gh/imega-docker/consul) [![GitHub stars](https://img.shields.io/github/stars/badges/shields.svg?style=social&label=Star&maxAge=2592000)](https://github.com/imega-docker/consul)

Image size: 33 MB

From image: alpine:3.4

Consul: 0.7.0-r0

## Usage

You need to create a fixture in the form of a directory tree. A value to be written to a file.

e.g. *http://consul:8500/v1/kv/service/mysql/host*

```
fixtures
 |- v1
    |- kv
        |- service
            |- mysql
                |- host
```

`mkdir -p fixtures/v1/kv/service/mysql/host`

`echo "my-host-mysql" >> fixtures/v1/kv/service/mysql/host/value.json`

### Run consul

```
docker run -d -v `pwd`/fixtures:/data -p 8500:8500 imega/consul
```

### Check value
```
$ curl http://consul:8500/v1/kv/service/mysql/host?raw
my-host-mysql
$
```

## Alpine Packages
  - musl (1.1.14-r11)
  - busybox (1.24.2-r11)
  - ncurses-terminfo-base (6.0-r7)
  - ncurses-terminfo (6.0-r7)
  - ncurses-libs (6.0-r7)
  - readline (6.3.008-r4)
  - bash (4.3.42-r3)
  - consul (0.7.0-r0)
  - zlib (1.2.8-r2)
  - libcrypto1.0 (1.0.2h-r2)
  - ca-certificates (20160104-r4)
  - libssl1.0 (1.0.2h-r2)
  - libssh2 (1.7.0-r0)
  - libcurl (7.50.2-r0)
  - curl (7.50.2-r0)
  - jq (1.5-r1)

## The MIT License (MIT)

Copyright © 2016 iMega <info@imega.ru>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.txt
### 0.7.0

  * Use base image on alpine 3.4
  * Consul 0.7.0-r0

### 0.5.2

  * Use base image on alpine 3.4
  * Consul 0.5.2-r1


tests/test.sh
#!/usr/bin/env bash

printf '%.0s-' {1..80}
echo

URL=$1

COUNT_TESTS=0
COUNT_TESTS_FAIL=0

assertTrue() {
    testName="$3"
    pad=$(printf '%0.1s' "."{1..80})
    padlength=78

    if [ "$1" != "$2" ]; then
        printf ' %s%*.*s%s' "$3" 0 $((padlength - ${#testName} - 4)) "$pad" "Fail"
        printf ' (expected %s, assertion %s)\n' "$1" "$2"
        let "COUNT_TESTS_FAIL++"
    else
        printf ' %s%*.*s%s\n' "$3" 0 $((padlength - ${#testName} - 2)) "$pad" "Ok"
        let "COUNT_TESTS++"
    fi
}

testDbname() {
    ACTUAL=$(curl --silent http://$URL/v1/kv/service/mysql/dbname?raw)

    assertTrue "db-name" $ACTUAL "$FUNCNAME"
}

testHost() {
    ACTUAL=$(curl --silent http://$URL/v1/kv/service/mysql/host?raw)

    assertTrue "mysql-server" $ACTUAL "$FUNCNAME"
}

testPassword() {
    ACTUAL=$(curl --silent http://$URL/v1/kv/service/mysql/password?raw)

    assertTrue "p@55w0rd" $ACTUAL "$FUNCNAME"
}

testPort() {
    ACTUAL=$(curl --silent http://$URL/v1/kv/service/mysql/port?raw)

    assertTrue "3306" $ACTUAL "$FUNCNAME"
}

testUser() {
    ACTUAL=$(curl --silent http://$URL/v1/kv/service/mysql/user?raw)

    assertTrue "root" $ACTUAL "$FUNCNAME"
}

while [[ "`curl -s http://$URL/v1/status/leader --write-out %{http_code} --silent --output /dev/null`" != "200" || "`curl -s http://$URL/v1/status/leader`" == "\"\"" ]]; do \
    sleep 0.3; \
done

sleep 1;

testDbname
testHost
testPassword
testPort
testUser

printf '%.0s-' {1..80}
echo
printf 'Total test: %s, fail: %s\n\n' "$COUNT_TESTS" "$COUNT_TESTS_FAIL"

if [ $COUNT_TESTS_FAIL -gt 0 ]; then
    exit 1
fi

exit 0


tests/wait.sh
#!/usr/bin/env bash

MOCK_SERVER_CONSUL=$1

while [[ "`curl -s http://$MOCK_SERVER_CONSUL/v1/status/leader --write-out %{http_code} --silent --output /dev/null`" != "200" || "`curl -s http://$MOCK_SERVER_CONSUL/v1/status/leader`" == "\"\"" ]]; do \
    echo "."; sleep 0.3; \
done


tests/import.sh
#!/usr/bin/env bash

MOCK_SERVER_CONSUL=$1

FIXTURES=$(find fixtures/consul -type f -path "*/v1/kv/*" -a -name *.json)
while IFS=';' read -ra ADDR; do
    for i in "${ADDR[@]}"; do
        DIRNAME=$(dirname $i)
        VALUE=$(cat "$i")
        curl -s -X PUT -d "$VALUE" http://$MOCK_SERVER_CONSUL${DIRNAME:15}
        echo " - "${DIRNAME:15}
    done
done <<< "$FIXTURES"

FIXTURES=$(find fixtures/consul -type f -path "*/v1/catalog/*" -a -name *.json)
while IFS=';' read -ra ADDR; do
    for i in "${ADDR[@]}"; do
        DIRNAME=$(dirname $i)
        VALUE=$(cat "$i" | jq -c .)
        curl -s -X PUT -d "$VALUE" http://$MOCK_SERVER_CONSUL${DIRNAME:15}
        echo " - "${DIRNAME:15}
    done
done <<< "$FIXTURES"


src/entry.sh
#!/usr/bin/env bash

MOCK_SERVER_CONSUL="127.0.0.1:8500"

function import() {
    FIXTURES=$(find /data -type f -path "*/v1/kv/*" -a -name *.json)
    while IFS=';' read -ra ADDR; do
        for i in "${ADDR[@]}"; do
            DIRNAME=$(dirname $i)
            VALUE=$(cat "$i")
            curl -s -X PUT -d "$VALUE" http://$MOCK_SERVER_CONSUL${DIRNAME:5}
            echo " - "${DIRNAME:5}
        done
    done <<< "$FIXTURES"

    FIXTURES=$(find /data -type f -path "*/v1/catalog/*" -a -name *.json)
    while IFS=';' read -ra ADDR; do
        for i in "${ADDR[@]}"; do
            DIRNAME=$(dirname $i)
            VALUE=$(cat "$i" | jq -c .)
            curl -s -X PUT -d "$VALUE" http://$MOCK_SERVER_CONSUL${DIRNAME:5}
            echo " - "${DIRNAME:5}
        done
    done <<< "$FIXTURES"
}

while `sleep 0.3;true`; do \
    if [[ `consul info | grep -e "leader = true" -e "state = Leader" | tr '\t\n' ' ' | sed 's/ //g'` == "leader=truestate=Leader" ]]; then \
        import; \
        break; \
    fi; \
done &

consul agent -client=0.0.0.0 -dev
